### PR TITLE
Fix/remove mocked acp api

### PR DESCRIPTION
**Pull request by megoth**
# Removing mocked ACP API

This removes the mocked API that we set up for ACP functionality. This is using a prerelease of the Solid Client, explicitly made to offer the new ACP API.

Still waiting for the release of 0.6.2, but everything else should be ok.

# Checklist

- [x] All acceptance criteria are met.
- [ ] Relevant documentation, if any, has been written/updated.
- [ ] The changelog has been updated, if applicable.
- [x] Commits in this PR are minimal and [have descriptive commit messages](https://chris.beams.io/posts/git-commit/).

### CODE REVIEW

**Review by ajacksified — COMMENTED** [review]
With the ESS demo, I get `Creating the empty Container at `https://ldp.demo-ess.inrupt.com/105169495685245623105/pb_policies/` failed: 403 Forbidden`, and a blank screen. It seems to work with the Gluu broker, though.

Also, the path to policies seems broken: it created `p/pb_policies/ublic` instead of `pb_policies/public`, although it seems to work anyways:

Screen Shot 2020-10-30 at 1 19 58 AM

Code looks good otherwise.

**Comment by megoth** [comment]
> With the ESS demo, I get `Creating the empty Container at `[https://ldp.demo-ess.inrupt.com/105169495685245623105/pb_policies/`](https://ldp.demo-ess.inrupt.com/105169495685245623105/pb_policies/%60) failed: 403 Forbidden`, and a blank screen. It seems to work with the Gluu broker, though.
> 
> Also, the path to policies seems broken: it created `p/pb_policies/ublic` instead of `pb_policies/public`, although it seems to work anyways:
> 
> Screen Shot 2020-10-30 at 1 19 58 AM
> 
> Code looks good otherwise. Will need to fix the path and make sure it's tested.

The problem with ESS demo probably is because the account was created before ACP was enabled. They are working on solving that. For now you should login with a unused OAuth account to create a new Pod.

**Review by ajacksified — APPROVED** [review]

File: src/accessControl/acp/index.js [hunk 1]
```diff
@@ -112,61 +90,88 @@ export function getOrCreatePermission(permissions, webId) {
...
-  return rules;
+  return { existing: true, rules };
 }
 
 export function getRuleWithAgent(rules, agentWebId) {
   // assumption 1: the rules for the policies we work with will not have agents across multiple rules
   // assumption 2: there will always be at least one rule (we enforce this with getRulesOrCreate)
   const rule = rules.find((r) =>
-    getAgentForRuleAll(r).find((webId) => webId === agentWebId)
+    acp.getAgentForRuleAll(r).find((webId) => webId === agentWebId)
```
Comment (NSeydoux) [inline src/accessControl/acp/index.js:129]: `acp.getAgentForRuleAll` would now be `acp.getAgentAll`

File: src/accessControl/acp/index.js [hunk 2]
```diff
@@ -112,61 +90,88 @@ export function getOrCreatePermission(permissions, webId) {
...
   );
   // if we don't find the agent in a rule, we'll just pick the first one
   return rule || rules[0];
 }
 
-export async function setAgents(policy, policyDataset, webId, accessToMode) {
-  const ruleUrls = getRequiredRuleOnPolicyAll(policy);
-  const rules = getRulesOrCreate(ruleUrls, policy, policyDataset);
+export function setAgents(policy, policyDataset, webId, accessToMode) {
+  const ruleUrls = acp.getRequiredRuleForPolicyAll(policy);
```
Comment (NSeydoux) [inline src/accessControl/acp/index.js:136]: `getRequiredRuleForPolicyAll` would now be `getRequiredRuleAll`

File: src/accessControl/acp/index.js [hunk 3]
```diff
@@ -112,61 +90,88 @@ export function getOrCreatePermission(permissions, webId) {
...
+  const { existing, rules } = getRulesOrCreate(ruleUrls, policy, policyDataset);
   const rule = getRuleWithAgent(rules, webId);
-  const existingAgents = getAgentForRuleAll(rule);
+  const existingAgents = acp.getAgentForRuleAll(rule);
```
Comment (NSeydoux) [inline src/accessControl/acp/index.js:139]: ditto

File: src/accessControl/acp/index.js [hunk 4]
```diff
@@ -112,61 +90,88 @@ export function getOrCreatePermission(permissions, webId) {
...
   const agentIndex = existingAgents.indexOf(webId);
-  let modifiedAgents;
+  let modifiedRule = rule;
   if (accessToMode && agentIndex === -1) {
-    modifiedAgents = existingAgents.concat(webId);
+    modifiedRule = acp.addAgentForRule(rule, webId);
```
Comment (NSeydoux) [inline src/accessControl/acp/index.js:143]: `addAgentForRule` -> `addAgent`

File: src/accessControl/acp/index.js [hunk 5]
```diff
@@ -112,61 +90,88 @@ export function getOrCreatePermission(permissions, webId) {
...
   }
   if (!accessToMode && agentIndex !== -1) {
-    modifiedAgents = existingAgents
-      .slice(0, agentIndex)
-      .concat(existingAgents.slice(agentIndex + 1));
+    modifiedRule = acp.removeAgentForRule(rule, webId);
```
Comment (NSeydoux) [inline src/accessControl/acp/index.js:146]: `removeAgentForRule` -> `removeAgent`

File: src/accessControl/acp/index.js [hunk 6]
```diff
@@ -112,61 +90,88 @@ export function getOrCreatePermission(permissions, webId) {
...
   }
-  const modifiedRule = setAgentInRule(rule, modifiedAgents);
-  return setRequiredRuleOnPolicy(policy, modifiedRule);
+  const modifiedDataset = setThing(policyDataset, modifiedRule);
+  return {
+    policy: existing
+      ? acp.setRequiredRuleForPolicy(policy, modifiedRule)
```
Comment (NSeydoux) [inline src/accessControl/acp/index.js:151]: `setRequiredRuleForPolicy` -> `setRequiredRule`

File: src/accessControl/acp/index.js [hunk 7]
```diff
@@ -112,61 +90,88 @@ export function getOrCreatePermission(permissions, webId) {
...
+      : acp.addRequiredRuleForPolicy(policy, modifiedRule),
```
Comment (NSeydoux) [inline src/accessControl/acp/index.js:152]: ditto

File: src/accessControl/acp/index.js [hunk 8]
```diff
@@ -112,61 +90,88 @@ export function getOrCreatePermission(permissions, webId) {
...
+    dataset: modifiedDataset,
+  };
+}
+
+export function getPolicyModesAndAgents(policyUrls, policyDataset) {
+  return policyUrls
+    .map((url) => acp.getPolicy(policyDataset, url))
+    .filter((policy) => !!policy)
+    .map((policy) => {
+      const modes = acp.getAllowModesOnPolicy(policy);
```
Comment (NSeydoux) [inline src/accessControl/acp/index.js:162]: `getAllowModesOnPolicy` -> `getAllowModes`

File: src/accessControl/acp/index.js [hunk 9]
```diff
@@ -112,61 +90,88 @@ export function getOrCreatePermission(permissions, webId) {
...
+      const ruleUrls = acp.getRequiredRuleForPolicyAll(policy);
```
Comment (NSeydoux) [inline src/accessControl/acp/index.js:163]: `getRequiredRuleForPolicyAll` -> `getRequiredRuleAll`

File: src/accessControl/acp/index.js [hunk 10]
```diff
@@ -112,61 +90,88 @@ export function getOrCreatePermission(permissions, webId) {
...
+      // assumption: rule resides in the same resource as policies
+      const rules = ruleUrls.map((url) => acp.getRule(policyDataset, url));
+      const agents = rules.reduce(
+        (memo, rule) => memo.concat(acp.getAgentForRuleAll(rule)),
```
Comment (NSeydoux) [inline src/accessControl/acp/index.js:167]: ditto

File: src/accessControl/acp/index.js [hunk 11]
```diff
@@ -255,33 +284,29 @@ export default class AcpAccessControlStrategy {
...
+      // no need to add policy if it doesn't exist and we're not adding the agent to it
+      return policyDataset;
+    }
     const { dataset: modifiedPolicyDataset } = chain(
       getOrCreatePolicy(policyDataset, policyUrl),
-      ({ policy: p, dataset: d }) => ({
-        policy: setAllowModesOnPolicy(p, acpMap),
-        dataset: d,
+      ({ policy, dataset }) => ({
+        policy: acp.setAllowModesOnPolicy(policy, acpMap),
```
Comment (NSeydoux) [inline src/accessControl/acp/index.js:296]: ditto

File: src/accessControl/acp/index.js [hunk 12]
```diff
@@ -291,23 +316,33 @@ export default class AcpAccessControlStrategy {
...
-      (p) => setAgents(p, policyDataset, webId, access.control)
+    const existingPolicy = acp.getPolicy(policyDataset, policyUrl);
+    if (!existingPolicy && !access.control) {
+      // no need to add policy if it doesn't exist and we're not adding the agent to it
+      return policyDataset;
+    }
+    const { dataset: modifiedPolicyDataset } = chain(
+      getOrCreatePolicy(policyDataset, policyUrl),
+      ({ policy, dataset }) => ({
+        policy: acp.setAllowModesOnPolicy(policy, createAcpMap(true, true)),
```
Comment (NSeydoux) [inline src/accessControl/acp/index.js:327]: ditto

File: src/accessControl/acp/index.js [hunk 13]
```diff
@@ -318,31 +353,30 @@ export default class AcpAccessControlStrategy {
...
+      { fetch: this.#fetch }
     );
-    const modifiedPolicyDataset = setPolicy(policyDataset, applyPolicy);
-    // making sure that policy is connected to access control
-    let accessControlWithPolicy = accessControls.find((ac) =>
-      getPolicyUrlAll(ac).find(policyUrl)
+    const { policy: modifiedPolicyDataset } = chain(
+      getOrCreatePolicy(policyDataset, policyUrl),
+      ({ policy, dataset }) => ({
+        policy: acp.setAllowModesOnPolicy(policy, createAcpMap(true, true)),
```
Comment (NSeydoux) [inline src/accessControl/acp/index.js:368]: ditto

File: src/accessControl/acp/index.test.js [hunk 14]
```diff
@@ -55,18 +83,283 @@ describe("AcpAccessControlStrategy", () => {
...
+      const error = new Error("500");
+      jest.spyOn(scFns, "getSolidDataset").mockImplementation(() => {
+        throw error;
+      });
+      await expect(acp.getPermissions()).rejects.toEqual(error);
+    });
+
+    it("normalizes the permissions retrieved from the policy resource", async () => {
+      const readPolicyRule = chain(acpFns.createRule(readPolicyRuleUrl), (r) =>
+        acpFns.addAgentForRule(r, webId)
```
Comment (NSeydoux) [inline src/accessControl/acp/index.test.js:163]: `addAgentForRule` -> `addAgent`

File: src/accessControl/acp/index.test.js [hunk 15]
```diff
@@ -55,18 +83,283 @@ describe("AcpAccessControlStrategy", () => {
...
+      );
+      const readPolicy = chain(
+        acpFns.createPolicy(readApplyPolicyUrl),
+        (p) => acpFns.setAllowModesOnPolicy(p, createAcpMap(true)),
```
Comment (NSeydoux) [inline src/accessControl/acp/index.test.js:167]: `setAllowModesOnPolicy` -> `setAllowedModes`

File: src/accessControl/acp/index.test.js [hunk 16]
```diff
@@ -55,18 +83,283 @@ describe("AcpAccessControlStrategy", () => {
...
+        (p) => acpFns.setRequiredRuleForPolicy(p, readPolicyRule)
```
Comment (NSeydoux) [inline src/accessControl/acp/index.test.js:168]: `setRequiredRuleForPolicy` -> `setRequiredRule`

File: src/accessControl/acp/index.test.js [hunk 17]
```diff
@@ -55,18 +83,283 @@ describe("AcpAccessControlStrategy", () => {
...
+      );
+      const controlPolicyRule = chain(
+        acpFns.createRule(controlPolicyRuleUrl),
+        (r) => acpFns.addAgentForRule(r, webId)
```
Comment (NSeydoux) [inline src/accessControl/acp/index.test.js:172]: ditto

File: src/accessControl/acp/index.test.js [hunk 18]
```diff
@@ -55,18 +83,283 @@ describe("AcpAccessControlStrategy", () => {
...
+      );
+      const controlPolicy = chain(
+        acpFns.createPolicy(controlAccessPolicyUrl),
+        (p) => acpFns.setAllowModesOnPolicy(p, createAcpMap(true, true)),
```
Comment (NSeydoux) [inline src/accessControl/acp/index.test.js:176]: ditto

File: src/accessControl/acp/index.test.js [hunk 19]
```diff
@@ -55,18 +83,283 @@ describe("AcpAccessControlStrategy", () => {
...
+        (p) => acpFns.setRequiredRuleForPolicy(p, controlPolicyRule)
```
Comment (NSeydoux) [inline src/accessControl/acp/index.test.js:177]: ditto

File: src/accessControl/acp/index.test.js [hunk 20]
```diff
@@ -178,58 +471,227 @@ describe("getOrCreatePermission", () => {
...
+    ).toEqual("http://example.com/pb_policies/pb_policies/test.ttl");
+    expect(
+      getPolicyUrl(mockSolidDatasetFrom(`${podUrl}public`), policies)
+    ).toEqual("http://example.com/pb_policies/public.ttl");
+  });
+});
+
+describe("getRulesOrCreate", () => {
+  const readPolicy = chain(acpFns.createPolicy(readApplyPolicyUrl), (p) =>
+    acpFns.setAllowModesOnPolicy(p, createAcpMap(true))
```
Comment (NSeydoux) [inline src/accessControl/acp/index.test.js:530]: ditto

File: src/accessControl/acp/index.test.js [hunk 21]
```diff
@@ -178,58 +471,227 @@ describe("getOrCreatePermission", () => {
...
+  });
+});
+
+describe("getRuleWithAgent", () => {
+  const ruleUrl = "http://example.com/#Rule";
+  const rule = acpFns.createRule(ruleUrl);
+  const webId = "http://example.com/profile/card#me";
+
+  it("returns a rule if it's connected to the agent", () => {
+    const ruleWithAgent = acpFns.addAgentForRule(rule, webId);
```
Comment (NSeydoux) [inline src/accessControl/acp/index.test.js:561]: `addAgentForRule` -> `addAgent`

File: src/accessControl/acp/index.test.js [hunk 22]
```diff
@@ -178,58 +471,227 @@ describe("getOrCreatePermission", () => {
...
+  });
+
+  it("returns first rule if no rule is connected to the agent", () => {
+    expect(getRuleWithAgent([rule], webId)).toBe(rule);
+  });
+});
+
+describe("setAgents", () => {
+  const readPolicy = chain(acpFns.createPolicy(readApplyPolicyUrl), (p) =>
+    acpFns.setAllowModesOnPolicy(p, createAcpMap(true))
```
Comment (NSeydoux) [inline src/accessControl/acp/index.test.js:572]: ditto

File: src/accessControl/acp/index.test.js [hunk 23]
```diff
@@ -178,58 +471,227 @@ describe("getOrCreatePermission", () => {
...
+  );
+  const readPolicyRule = acpFns.createRule(readPolicyRuleUrl);
+  const policyDataset = chain(mockSolidDatasetFrom(policyResourceUrl), (d) =>
+    acpFns.setPolicy(d, readPolicy)
+  );
+  const webId = "http://example.com/profile/card#me";
+
+  describe("adding agent", () => {
+    it("will add new rule and add agent to it", () => {
+      const expectedRule = acpFns.addAgentForRule(readPolicyRule, webId);
```
Comment (NSeydoux) [inline src/accessControl/acp/index.test.js:582]: ditto

File: src/accessControl/acp/index.test.js [hunk 24]
```diff
@@ -178,58 +471,227 @@ describe("getOrCreatePermission", () => {
...
+      const expectedDataset = setThing(policyDataset, expectedRule);
+      const expectedPolicy = acpFns.addRequiredRuleForPolicy(
```
Comment (NSeydoux) [inline src/accessControl/acp/index.test.js:584]: ditto

File: src/accessControl/acp/index.test.js [hunk 25]
```diff
@@ -178,58 +471,227 @@ describe("getOrCreatePermission", () => {
...
+      );
+      expect(setAgents(readPolicy, policyDataset, webId, true)).toEqual({
+        policy: expectedPolicy,
+        dataset: expectedDataset,
+      });
+    });
+
+    it("will use existing rule and add agent to it", () => {
+      const policyDatasetWithRule = setThing(policyDataset, readPolicyRule);
+      const expectedRule = acpFns.addAgentForRule(readPolicyRule, webId);
```
Comment (NSeydoux) [inline src/accessControl/acp/index.test.js:596]: ditto

File: src/accessControl/acp/index.test.js [hunk 26]
```diff
@@ -178,58 +471,227 @@ describe("getOrCreatePermission", () => {
...
+      const expectedDataset = setThing(policyDatasetWithRule, expectedRule);
+      const expectedPolicy = acpFns.setRequiredRuleForPolicy(
```
Comment (NSeydoux) [inline src/accessControl/acp/index.test.js:598]: ditto

File: src/accessControl/acp/index.test.js [hunk 27]
```diff
@@ -178,58 +471,227 @@ describe("getOrCreatePermission", () => {
...
+      expect(setAgents(readPolicy, policyDatasetWithRule, webId, true)).toEqual(
+        {
+          policy: expectedPolicy,
+          dataset: expectedDataset,
+        }
+      );
+    });
+
+    it("will use existing rule and not add agent to it if agent already is added", () => {
+      const ruleWithAgent = acpFns.addAgentForRule(readPolicyRule, webId);
```
Comment (NSeydoux) [inline src/accessControl/acp/index.test.js:611]: `addAgentForRule` -> `addAgent`

File: src/accessControl/acp/index.test.js [hunk 28]
```diff
@@ -178,58 +471,227 @@ describe("getOrCreatePermission", () => {
...
+      const policyDatasetWithRule = setThing(policyDataset, ruleWithAgent);
+      const expectedDataset = setThing(policyDatasetWithRule, ruleWithAgent);
+      const expectedPolicy = acpFns.setRequiredRuleForPolicy(
```
Comment (NSeydoux) [inline src/accessControl/acp/index.test.js:614]: ditto

File: src/accessControl/acp/index.test.js [hunk 29]
```diff
@@ -178,58 +471,227 @@ describe("getOrCreatePermission", () => {
...
+          policy: expectedPolicy,
+          dataset: expectedDataset,
+        }
+      );
+    });
+  });
+
+  describe("removing agent", () => {
+    it("adds rule to policy and dataset if trying to remove agent when there are no rules yet", () => {
+      const expectedPolicy = acpFns.addRequiredRuleForPolicy(
```
Comment (NSeydoux) [inline src/accessControl/acp/index.test.js:629]: ditto

File: src/accessControl/acp/index.test.js [hunk 30]
```diff
@@ -178,58 +471,227 @@ describe("getOrCreatePermission", () => {
...
+      );
+      const expectedDataset = setThing(policyDataset, readPolicyRule);
+      expect(setAgents(readPolicy, policyDataset, webId, false)).toEqual({
+        policy: expectedPolicy,
+        dataset: expectedDataset,
+      });
+    });
+
+    it("removes agent from rule", () => {
+      const ruleWithAgent = acpFns.addAgentForRule(readPolicyRule, webId);
```
Comment (NSeydoux) [inline src/accessControl/acp/index.test.js:641]: ditto

File: src/accessControl/acp/index.test.js [hunk 31]
```diff
@@ -178,58 +471,227 @@ describe("getOrCreatePermission", () => {
...
+      const policyDatasetWithRule = setThing(policyDataset, ruleWithAgent);
+      const expectedDataset = setThing(policyDatasetWithRule, readPolicyRule);
+      const expectedPolicy = acpFns.setRequiredRuleForPolicy(
```
Comment (NSeydoux) [inline src/accessControl/acp/index.test.js:644]: ditto

File: src/accessControl/acp/index.test.js [hunk 32]
```diff
@@ -178,58 +471,227 @@ describe("getOrCreatePermission", () => {
...
+});
+
+describe("getPolicyModesAndAgents", () => {
+  it("maps agents from policies together with the access modes they're granted", () => {
+    const webId1 = "http://example.com/agent1";
+    const webId2 = "http://example.com/agent2";
+    const readPolicyRule = chain(
+      acpFns.createRule(readPolicyRuleUrl),
+      (r) => acpFns.addAgentForRule(r, webId1),
+      (r) => acpFns.addAgentForRule(r, webId2)
```
Comment (NSeydoux) [inline src/accessControl/acp/index.test.js:665]: ditto

File: src/accessControl/acp/index.test.js [hunk 33]
```diff
@@ -178,58 +471,227 @@ describe("getOrCreatePermission", () => {
...
+    );
+    const readPolicy = chain(
+      acpFns.createPolicy(readApplyPolicyUrl),
+      (p) => acpFns.setAllowModesOnPolicy(p, createAcpMap(true)),
+      (p) => acpFns.setRequiredRuleForPolicy(p, readPolicyRule)
```
Comment (NSeydoux) [inline src/accessControl/acp/index.test.js:670]: ditto

File: src/accessControl/acp/index.test.js [hunk 34]
```diff
@@ -178,58 +471,227 @@ describe("getOrCreatePermission", () => {
...
+    );
+    const writePolicyRule = chain(acpFns.createRule(writePolicyRuleUrl), (r) =>
+      acpFns.addAgentForRule(r, webId2)
```
Comment (NSeydoux) [inline src/accessControl/acp/index.test.js:673]: ditto

File: src/accessControl/acp/index.test.js [hunk 35]
```diff
@@ -178,58 +471,227 @@ describe("getOrCreatePermission", () => {
...
+    );
+    const writePolicy = chain(
+      acpFns.createPolicy(writeApplyPolicyUrl),
+      (p) => acpFns.setAllowModesOnPolicy(p, createAcpMap(false, true)),
+      (p) => acpFns.setRequiredRuleForPolicy(p, writePolicyRule)
```
Comment (NSeydoux) [inline src/accessControl/acp/index.test.js:678]: ditto

**Review by NSeydoux — COMMENTED** [review]
Just a quick search and replace to update to the latest API :)

**Pull request merged**